### DISCUSSION TITLE
Misalignment of components on back ground image

### DISCUSSION THREAD

**Discussion by bijlevel** [RomanKubiak/ctrlr#334] [Q&A] [answered]
As I mentioned a few times in [the keypress discussion](https://github.com/RomanKubiak/ctrlr/discussions/256), I have an issue with components ending up in different positions after exporting the panel as .bpanelz. When I save and load the panel as .panel then everything is fine and components stay at the their position. This is what it looks like:

<img width="810" alt="Screenshot 2021-07-12 at 10 41 33" src="https://user-images.githubusercontent.com/11859363/125262434-73f7ee00-e302-11eb-9ba6-2011a9cc8fae.png">

The selected components are at the position I want them to be and stay there after a panel save and reload.

This is how it looks like after this panel is exported as .bpanelz:

<img width="808" alt="Screenshot 2021-07-12 at 10 43 02" src="https://user-images.githubusercontent.com/11859363/125262881-d4872b00-e302-11eb-8285-b463d80ab319.png">

When I load this panel, all the selected components are shifted to another position and the canvas (is it the canvas?) has become wider. I made sure the uiPanelZoom is **1.0**. On the old forum I did found a thread remotely linked to this problem and it was suggested to set the initial zoom from Lua in an 'initialisePanel' method. I experimented with this, but the panel does not pick up the value set in the method. So ANY input on this issue is greatly appreciated!

**Comment by dnaldoog** [comment, marked as answer]
Hi @bijlevel ,

I have had issues like this I think, but a solution that always works for me is to put the background image into a _uiComponent_ that sits in a layer below all the other visual elements. Of course set the _uiComponent_ dimensions to exactly that of the image.

<pre>
paintBackground = function(--[[ CtrlrComponent --]] comp, --[[ CtrlrComponent --]] g)
local bgResource = resources:getResource("BG")
        if bgResource ~= nil then
           local  bImage = bgResource:asImage()
            g:drawImageAt(bImage, 0, 0, false)
        end -- image found
end
</pre>


I suppose you already know the code for setting zoom to 100% is `    panel:getPanelEditor():setProperty("uiPanelZoom", 1, false)`.

**Reply by bijlevel** [reply]
@dnaldoog Thanks for this suggestion! I applied it and to very odd thing is that each time I export the same panel as .bpanelz the height (the y part of the Canvas bounds setting) is diminished by 24 (pixels?). The height of the image is 274 pixels, so after 11 times saving there is almost no height anymore! ;-).

BTW why should I use a zoom setting of zero?

**Reply by Ted-juh** [reply]
I think he meant 1, as in 100%. The 24 pixel problem is an ancient CTRLR bug. Add 24 pixels to the y-axis of the canvas bounds on panel loaded.

**Reply by bijlevel** [reply]
I remember now coming across this bug, but because I did not use images on uiCustomComponents in my earlier panels the way I'm doing it now, I never paid attention to this bug. So I will set the required canvas bounds+24 on panel load. Thanks for this advice!

**Reply by damiensellier** [reply]
Your GUI is very nice, I would suggest you to check the reflexion of the light on your big dial. It should stay at the same position even if the button is moving. The lights always stay at the same position in the room anyway. 
So you should have 2 layers in your knobman app. 1 with the knob including the bevel with the top reflexion without the recessed dimple FIXED.  Another with the recessed dimple off-axis from the button center which rotates. Its reflexion should stay at the same angle as well whatever the rotation it is( angle value from its center is opposite of the angle of the overall rotation).  This way you'll get something realistic.
I also tend do dislike plugins with screws, jack inputs etc they clutter the interface for nothing. Is the disk drive mockup linked to a feature like Load/save? is the jack linked to an input/output setup? The right side extension ports are functionnal?  if no, maybe you should think about removing it to gain real estate and legibility. But this is how I see it, just my general opinion on plugins GUI inspired by hardware. 
Nice job!

**Reply by dnaldoog** [reply]
> I think he meant 1, as in 100%. The 24 pixel problem is an ancient CTRLR bug. Add 24 pixels to the y-axis of the canvas bounds on panel loaded.

Okay - I changed that!

Also on any of my panels for example [this one](https://ctrlr.org/123352-2/), I have three buttons: one to zoom out, another to zoom in and another to reset to 100%. Please feel free to use that code! :)

**Reply by bijlevel** [reply]
@dnaldoog I did set the zoom using Lua, but this doesn't work in the versions I'm using (mac 5.3.122/5.3.198, win 5.3.201). I did try several values, but the panel never changed the zoom (after loading).

@damiensellier You're right about the reflection of the dial. But this isn't a Knobman knob, but an single image added to the uiCustomComponent. So making the refection realistic would be very nice, but ATM not our main concern. OTH the dial springs quite fast back to position zero, so the user won't hardly notice the reflection anomaly...
You made me think about the functionality of the panel background "items" (screws, jacks etc.) not used in the interface. We will give it a thought. Thanks for bringing this to our attention!

Next issue is now that the transparency of the dial has gone as soon as it is positioned above the background uiCustomComponent. It looks like this:

<img width="216" alt="Screenshot 2021-07-13 at 10 45 22" src="https://user-images.githubusercontent.com/11859363/125421207-a5a37486-18cd-4cc0-82f8-cf0536d59769.png">

This was never an issue when this component was placed above a background image. Tranparency has, I believe, to do with the so called alpha channel of an image. Is there a way to set the value of such a channel in CTRLR or is there an easier way?

**Reply by bijlevel** [reply]
Well, CTRLR at its best again ;-). I did modify another panel on which the dial was shown OK, by replacing the existing background image of that panel with a uiCustomComponent having this background image, and now the transparency is back again.

**Reply by Ted-juh** [reply]
From the top of my head, is there a g:setOpacity() after you imported the image?

The lightning on the dial can be done. The big dial could be a static picture as the background of the component. The indent (recessed dimple, a separate picture) could spin a certain amount from the centre. And to keep the lightning at the right angle, rotate the recessed dimple itself 6.28 radians ( or 2 * pi) for a whole spin. But that's not high priority right now.

**Reply by damiensellier** [reply]
PNG with alpha channel bug in certain versions of ctrlr mostly with uiButtonImage. When you play with those component the part of the image where there's a bit of transparency stacks up on every move, and in the end you have big blob of color :) I alway use PNG without transparency with the background included.
Since you already have -/+ buttons down the dial you can  even remove it, you will get even more space and won't lose this feature since it's a duplicate. If I were you I would put the -/+ buttons above or below the navigation buttons up/dwn/lft/right and nudge left the alpha keyboard, and remove the disk drive and decorative right hand side ports.

**Reply by bijlevel** [reply]
@DnaldeT yes, there is a g:setOpacity(1.0) statement, but changing the values doesn't have an effect on the component. But as I said, with my current panel all is well!

Thanks for explaining the way to handle the lighting of the dial, I'm sure I will need it sometime in the near future, but as I wrote, this is ATM not high on my priority list.

@damiensellier I did not have issues before with PNG's and transparency, but again, at the moment all is well.

**Reply by bijlevel** [reply]
> Also on any of my panels for example [this one](https://ctrlr.org/123352-2/), I have three buttons: one to zoom out, another to zoom in and another to reset to 100%. Please feel free to use that code! :)

John,
I did download the .bpanelz version, but I can't access the code because the menubar is not there... I was hoping to find also the right coding for the uiPanelCanvasRectangle because the way I did it, adding 

`panel:getPanelEditor():setProperty("uiPanelZoom",0.6,true)
panel:getPanelEditor():setProperty("uiPanelCanvasRectangle","0 0 1818 800",true)`

to the initialise method does not have any effect on the size of the panel. Oddly enough it keeps the size of the background uiCustomComponent, the one with the background image (size 1818x553).

**Reply by Ted-juh** [reply]
You can get access to the menubar by pressing on the orange triangle in the left corner.  But then you cannot choose the panel menu to get to the editor. Better try his JV-1010 panel. https://ctrlr.org/jv-1010-editor/ 

The functions for zooming are in the "zoomPanel" script under "various".

You add the 24 pixels to the canvas bounds before exporting the panel as VST, Standalone, or bpanelz.  Sorry, my earlier comment suggested you had to do it in panel loaded, which was a wrong assumption.

**Reply by bijlevel** [reply]
@DnaldeT Thanks again! So I have to remember to add 24 pix to the vertical (y-axis) value of the Canvas Rectangle each time I do an export. BTW I did test this in the latest version (5.6.23) and the bug seems gone. When opening a .bpanelz panel, the Canvas Rectangle keeps its values.
Question for @dnaldoog how do you create a .bpanelz file without the menubar of CTRLR?

**Comment by dnaldoog** [comment]
Hi @bijlevel!

Here is how I do it:

`bit.bxor(value,1)` changes a 0 into a 1 and a 1 into a 0!

<pre>
hideShowMenu = function(mod,value,source)
    local status = panel:getPanelEditor():getPropertyInt("uiPanelMenuBarVisible")
    local b = flipZeroWithOne(status)
    panel:getPanelEditor():setProperty("uiPanelMenuBarVisible", b, false)
end --function

flipZeroWithOne = function(n)
    return  bit.bxor(n, 1)
end --function
</pre>

----

[hide menu_1.0.panel.zip](https://github.com/RomanKubiak/ctrlr/files/6826426/hide.menu_1.0.panel.zip)

**Reply by bijlevel** [reply]
@dnaldoog Thanks a lot, much appreciated! Amazing how you know how to code these things. Changing properties I understand these days, but the "variant" 'getPropertyInt' (adding "Int") is new for me, as is the way to flip a zero with a one!

**Reply by dnaldoog** [reply]
I just mostly copy and plagiarise! That's all! :)

You can access a list of most of the component properties you want to modify  this way:

![components](https://user-images.githubusercontent.com/76831683/125929953-8a3ac9d7-e7dc-410f-84e0-17436d5bca30.png)
----
There is for example:
`local x="LFO_Rate"`

`panel:getModulatorByName(x):getComponent():setProperty("componentLabelColour","FF000000",true)`

There is also `setPropertyInt` and `setPropertyString`. I don't think they take the extra _boolean_ argument.

and sometimes you need to _get_ a property:

`panel:getModulatorByName(x):getComponent():getProperty("componentLabelColour")`

Did you know you can create your own __custom property__?

`panel:getModulatorByName(x):getComponent():setProperty("myCustomIndex",77,true)`

and if you want to remove it:
`panel:getModulatorByName(x):getComponent():removeProperty("myCustomIndex")`

:)

**Reply by bijlevel** [reply]
@dnaldoog very handy! Thanks for showing me/us.
